common/nymcoconut/src/scheme/mod.rs

```diff
@@ -0,0 +1,600 @@
+        let h_prime = self.0 * r_prime;
+        let s_prime = (self.1 * r_prime) + (h_prime * r);
+        (Signature(h_prime, s_prime), r)
+    }
+
+    pub fn to_bytes(self) -> [u8; 96] {
+        let mut bytes = [0u8; 96];
+        bytes[..48].copy_from_slice(&self.0.to_affine().to_compressed());
+        bytes[48..].copy_from_slice(&self.1.to_affine().to_compressed());
+        bytes
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Result<Signature> {
+        Signature::try_from(bytes)
+    }
+}
+
+impl Bytable for Signature {
+    fn to_byte_vec(&self) -> Vec<u8> {
+        self.to_bytes().to_vec()
+    }
+
+    fn try_from_byte_slice(slice: &[u8]) -> Result<Self> {
+        Signature::from_bytes(slice)
+    }
+}
+
+impl Base58 for Signature {}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct BlindedSignature(G1Projective, elgamal::Ciphertext);
+
+impl Bytable for BlindedSignature {
+    fn to_byte_vec(&self) -> Vec<u8> {
+        self.to_bytes().to_vec()
+    }
+
+    fn try_from_byte_slice(slice: &[u8]) -> Result<Self> {
+        Self::from_bytes(slice)
+    }
+}
+
+impl Base58 for BlindedSignature {}
+
+impl TryFrom<&[u8]> for BlindedSignature {
+    type Error = CoconutError;
+
+    fn try_from(bytes: &[u8]) -> Result<BlindedSignature> {
+        if bytes.len() != 144 {
+            return Err(CoconutError::Deserialization(format!(
+                "BlindedSignature must be exactly 144 bytes, got {}",
+                bytes.len()
+            )));
+        }
+
+        let h_bytes: &[u8; 48] = &bytes[..48].try_into().expect("Slice size != 48");
+
+        let h = try_deserialize_g1_projective(
+            h_bytes,
+            CoconutError::Deserialization("Failed to deserialize compressed h".to_string()),
+        )?;
+        let c_tilde = Ciphertext::try_from(&bytes[48..])?;
+
+        Ok(BlindedSignature(h, c_tilde))
+    }
+}
+
+impl BlindedSignature {
+    pub fn unblind(
+        &self,
+        params: &Parameters,
+        private_key: &elgamal::PrivateKey,
+        partial_verification_key: &VerificationKey,
+        private_attributes: &[Attribute],
+        public_attributes: &[Attribute],
+        commitment_hash: &G1Projective,
+    ) -> Result<Signature> {
+        // parse the signature
+        let h = &self.0;
+        let c = &self.1;
+        let sig2 = private_key.decrypt(c);
+
+        // Verify the commitment hash
+        if !(commitment_hash == h) {
+            return Err(CoconutError::Unblind(
+                "Verification of commitment hash from signature failed".to_string(),
+            ));
+        }
+
+        let alpha = partial_verification_key.alpha;
+
+        let tmp = private_attributes
+            .iter()
+            .chain(public_attributes.iter())
+            .zip(partial_verification_key.beta.iter())
+            .map(|(attr, beta_i)| beta_i * attr)
+            .sum::<G2Projective>();
+
+        // Verify the signature share
+        if !check_bilinear_pairing(
```

Comment:
Do we really have to perform the pairing for each unblind operation?

Reply:
Yes, we have to make sure that what we got back from the validator is legit